### PR TITLE
Adds timeout and waits for feedback when asking the leader loop to reassert.

### PR DESCRIPTION
**Pull request by slackpad**
This adds on to the fix in #3004 for issue #2980.

### CODE REVIEW

File: consul/snapshot_endpoint.go [hunk 1]
```diff
@@ -101,10 +101,34 @@ func (s *Server) dispatchSnapshotRequest(args *structs.SnapshotRequest, in io.Re
...
+		timeoutCh := time.After(time.Minute)
+
 		select {
 		// Tell the leader loop to reassert leader actions since we just
 		// replaced the state store contents.
-		case s.reassertLeaderCh <- struct{}{}:
+		case s.reassertLeaderCh <- errCh:
+
+		// We might have lost leadership while waiting to kick the loop.
+		case <-timeoutCh:
```
Comment (magiconair) [inline consul/snapshot_endpoint.go]: I think you can inline `time.After()` here

File: consul/leader.go [hunk 2]
```diff
@@ -76,6 +76,19 @@ func (s *Server) leaderLoop(stopCh chan struct{}) {
 	var reconcileCh chan serf.Member
 	establishedLeader := false
 
+	reassert := func() error {
```
Comment (magiconair) [inline consul/leader.go]: why move this away? I'd keep the logic where it is used unless you want to call `reassert` multiple times. Easier to read.

Comment (slackpad) [inline consul/leader.go]: I thought this ended up cleaner than having an error write to the channel and a `continue` on each line of the function - made the `select` simpler.

File: consul/snapshot_endpoint.go (hunk 1 above)
Comment (slackpad) [inline consul/snapshot_endpoint.go]: Was thinking an overall minute for queue and run vs. restarting another minute here. It's pretty arbitrary though.

**Pull request merged**